test/request-timeout.js

```diff
@@ -0,0 +1,391 @@
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 50 }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 50 }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 50 }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+  })
+})
+
+test('Global option', (t) => {
+  t.plan(1)
+
+  const clock = FakeTimers.install()
+  t.teardown(clock.uninstall.bind(clock))
+
+  const server = createServer((req, res) => {
+    setTimeout(() => {
+      res.end('hello')
+    }, 100)
+    clock.tick(100)
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`, { requestTimeout: 50 })
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET' }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+
+    clock.tick(50)
+  })
+})
+
+test('Request options overrides global option', (t) => {
+  t.plan(1)
+
+  const clock = FakeTimers.install()
+  t.teardown(clock.uninstall.bind(clock))
+
+  const server = createServer((req, res) => {
+    setTimeout(() => {
+      res.end('hello')
+    }, 100)
+    clock.tick(100)
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`, { requestTimeout: 100 })
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 50 }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+
+    clock.tick(50)
+  })
+})
+
+test('client.destroy should cancel the timeout', (t) => {
+  t.plan(2)
+
+  const server = createServer((req, res) => {
+    res.end('hello')
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 100 }, (err, response) => {
+      t.ok(err instanceof errors.ClientDestroyedError)
+    })
+
+    client.destroy(err => {
+      t.error(err)
+    })
+  })
+})
+
+test('client.close should wait for the timeout', (t) => {
+  t.plan(2)
+
+  const clock = FakeTimers.install()
+  t.teardown(clock.uninstall.bind(clock))
+
+  const server = createServer((req, res) => {
+  })
+  t.teardown(server.close.bind(server))
+
+  server.listen(0, () => {
+    const client = new Client(`http://localhost:${server.address().port}`)
+    t.teardown(client.destroy.bind(client))
+
+    client.request({ path: '/', method: 'GET', requestTimeout: 100 }, (err, response) => {
+      t.ok(err instanceof errors.RequestTimeoutError)
+    })
+
+    client.close((err) => {
+      t.error(err)
```

Comment:
@mcollina @delvedor One problem... `RequestTimeoutError`s do not propagate to `close`. There is no elegant way to make it so with the current code... is this a problem?

Only "hard" errors that destroy the socket will propagate to `close` at the moment.

Reply:
The error modes of `close` are a bit ambiguous at the moment.

Reply:
I like the current solution, it makes sense that `destroy` closes everything as soon as you call it, while `close` uses a graceful shutdown :)

Reply:
With the current solution requests can still timeout after close and those errors are not propagated to close... which is the problem, or not? needs to be documented at least

Reply:
I don't see it as a problem. @mcollina?

Reply:
I don't understand this. Can you please recap it for me?

Reply:
So `close(err)` will be invoked with an error if any request fails... unless the request errors due to an error from abort or timeout.

Reply:
Because of this I'm kind of considering removing `close` until we figure out what the exact semantics of it is...

One definition could be that `client.close()` should be equal to `Promise.all(client.allRequestsAsPromises)`? Would require a little work to make that happen.

Reply:
I will fix this as part of https://github.com/mcollina/undici/issues/175 after this PR has landed. So this is no longer blocking.